src/hotspot/share/c1/c1_GraphBuilder.cpp

```diff
@@ -1741,15 +1741,17 @@ void GraphBuilder::access_field(Bytecodes::Code code) {
     }
   }
 
-  if (field->is_final() && (code == Bytecodes::_putfield)) {
-    scope()->set_wrote_final();
-  }
-
   if (code == Bytecodes::_putfield) {
     scope()->set_wrote_fields();
     if (field->is_volatile()) {
       scope()->set_wrote_volatile();
     }
+    if (field->is_final()) {
+      scope()->set_wrote_final();
+    }
+    if (field->is_stable()) {
```

Comment:
What if the `field` is a field of another object, not the one in construction? For final fields the verifier ensures that we only write to it in the containing object constructor, but for final fields it is not guaranteed.

Reply:
Isn't membar additions already checked with `is_object_initializer()` or `method()->name() == ciSymbols::object_initializer_name()`?

Reply:
Yes we only emit membars at the end of constructors but we do not check if the stable fields being written are the same objects as the ones being constructed there.

Reply:
That should be still fine, as this shouldn't affect performance-sensitive fast paths.